augur/merge.py

```diff
@@ -0,0 +1,304 @@
+            newline = os.linesep
+
+            print_info(f"Reading {m.name!r} metadata from {m.path!r}…")
+            sqlite3(db_path,
+                f'.mode csv',
+                f'.separator {sqlite_quote_dot(m.delimiter)} {sqlite_quote_dot(newline)}',
+                f'.import {sqlite_quote_dot(f"|{augur} read-file {shquote(m.path)}")} {sqlite_quote_dot(m.table_name)}',
+
+                f'create unique index {sqlite_quote_id(f"{m.table_name}_id")} on {sqlite_quote_id(m.table_name)}({sqlite_quote_id(m.id_column)});',
+
+                # <https://sqlite.org/pragma.html#pragma_optimize>
+                f'pragma optimize;')
+
+            # We're going to use Metadata.columns to generate the select
+            # statement, so ensure it matches what SQLite's .import created.
+            assert m.columns == (table_columns := sqlite3_table_columns(db_path, m.table_name)), \
+                f"{m.columns!r} == {table_columns!r}"
+
+            # Track which columns appear in which metadata inputs, preserving
+            # the order of both.
+            for column in m.columns:
+                # Match different id column names in different metadata files
+                # since they're logically equivalent.
+                output_column = output_id_column if column == m.id_column else column
+
+                output_columns.setdefault(output_column, [])
+                output_columns[output_column] += [(m.table_name, column)]
+
+
+        # Construct query to produce merged metadata.
+        select_list = [
+            # Output metadata columns coalesced across input metadata columns
+            *(f"""coalesce({', '.join(f"nullif({x}, '')" for x in starmap(sqlite_quote_id, reversed(input_columns)))}, null) as {sqlite_quote_id(output_column)}"""
+                for output_column, input_columns in output_columns.items()),
+
+            # Source columns
+            *(f"""{sqlite_quote_id(m.table_name, m.id_column)} is not null as {sqlite_quote_id(f'__source_metadata_{m.name}')}"""
+                for m in metadata)]
+
+        from_list = [
+            sqlite_quote_id(metadata[0].table_name),
+            *(f"full outer join {sqlite_quote_id(m.table_name)} on {sqlite_quote_id(m.table_name, m.id_column)} in ({', '.join(sqlite_quote_id(m.table_name, m.id_column) for m in reversed(preceding))})"
+                for m, preceding in [(m, metadata[:i]) for i, m in enumerate(metadata[1:], 1)])]
+
+        # Take some small pains to make the query readable since it makes
+        # debugging and development easier.  Note that backslashes aren't
+        # allowed inside f-string expressions, hence the *newline* variable.
+        newline = '\n'
+        query = dedent(f"""\
+            select
+                {(',' + newline + '                ').join(select_list)}
+            from
+                {(newline + '                ').join(from_list)}
+            ;
+            """)
```

Comment:
In case anyone else is wondering, this is what the query looks like for the 3-file example from tests:

```sql
select
    coalesce(nullif("metadata_z"."strain", ''), nullif("metadata_y"."strain", ''), nullif("metadata_x"."strain", ''), null) as "strain",
    coalesce(nullif("metadata_x"."a", ''), null) as "a",
    coalesce(nullif("metadata_y"."b", ''), nullif("metadata_x"."b", ''), null) as "b",
    coalesce(nullif("metadata_z"."c", ''), nullif("metadata_y"."c", ''), nullif("metadata_x"."c", ''), null) as "c",
    coalesce(nullif("metadata_y"."f", ''), null) as "f",
    coalesce(nullif("metadata_y"."e", ''), null) as "e",
    coalesce(nullif("metadata_y"."d", ''), null) as "d",
    coalesce(nullif("metadata_z"."g", ''), null) as "g",
    "metadata_x"."strain" is not null as "__source_metadata_x",
    "metadata_y"."strain" is not null as "__source_metadata_y",
    "metadata_z"."strain" is not null as "__source_metadata_z"
from
    "metadata_x"
    full outer join "metadata_y" on "metadata_y"."strain" in ("metadata_x"."strain")
    full outer join "metadata_z" on "metadata_z"."strain" in ("metadata_y"."strain", "metadata_x"."strain")
;
```

Reply:
I'm going to implement #1308 to print this sort of stuff out more easily.

Reply:
https://github.com/nextstrain/augur/pull/1563/commits/67551f1f26e52d9be886d0ddf664dfc198b0966f and https://github.com/nextstrain/augur/pull/1563/commits/30e8eb6a69b3c6f47f07495d4cfd9cdcb5ba57c0 do this.

Reply:
Moved the former commit into #1577 since it's a separate concern we want to merge independent of this, and [squashed](https://tsibley.net/paste/2024-08-14CHalJrOK-.html) the latter commit into this PR's commit.